cpp/src/arrow/util/mutex.cc

```diff
@@ -50,5 +55,31 @@ Mutex::Guard Mutex::Lock() {
 
 Mutex::Mutex() : impl_(new Impl, [](Impl* impl) { delete impl; }) {}
 
+#ifndef _WIN32
+namespace {
+
+struct AfterForkState {
+  // A global instance that will also register the atfork handler when
+  // constructed.
+  static AfterForkState instance;
+
+  // The mutex may be used at shutdown, so make it eternal.
+  // The leak (only in child processes) is a small price to pay for robustness.
+  Mutex* mutex = nullptr;
```

Comment:
nit: Maybe it's not trivial when working with multiprocessing in Python. Why not use a `unique_ptr` here?

Reply:
There isn't a lot of control over shutdown ordering at the moment.  So it's possible the main thread starts tearing down and deleting any unique_ptr here before the child threads have fully shutdown.  Then the child threads might still try and access this mutex which would lead to accessing freed memory.

Reply:
> Maybe it's not trivial when working with multiprocessing in Python

Do you have a particular use case in mind?  For example, the most common multiprocessing pattern is "fork-join" where you fork off a number of child processes for some task and then join them all back together when the task is done.  In that case the leak would be limited because the memory would be free when the child process ended.

The only way to really grow this leak is to recursively and continuously fork.  Even then you would run out of process IDs far before any significant memory was lost.  I suppose you could be continuously killing the parent off after the fork and leaving an orphan but I don't know of any real world use case for this.

Reply:
> There isn't a lot of control over shutdown ordering at the moment. So it's possible the main thread starts tearing down and deleting any unique_ptr here before the child threads have fully shutdown. Then the child threads might still try and access this mutex which would lead to accessing freed memory.

I see, thanks.

> Do you have a particular use case in mind?

My intuition is that one may create and destroy the`multiprocessing.Pool` continuously, which may result in too many leaks here.

Reply:
> My intuition is that one may create and destroy themultiprocessing.Pool continuously, which may result in too many leaks here.

That should be ok.  For example, imagine your process is on pid 100.   You create `multiprocessing.Pool` with size 10.  It will create processes 101, 102, ..., 110.  Each of those child processes will be just a tiny bit too large (40 bytes).  Process 100 will still be the correct size.  When you destroy the pool the processes 101-110 go away.  Process 100 remains and is still the correct size.  When you create a new pool the whole thing repeats with 10 new process IDs.

Reply:
Just for the record, no process will be larger because of this leak. Perhaps I should have put "leak" in quotes, because regardless of using `unique_ptr` or not, the mutex will be destroyed at process exit. The difference is mainly in code hygiene, and also some sanitizers may report it as an actual leak (memory that wasn't deallocated before process exit).

Reply:
> That should be ok. For example, imagine your process is on pid 100. You create `multiprocessing.Pool` with size 10. It will create processes 101, 102, ..., 110. Each of those child processes will be just a tiny bit too large (~40 bytes). Process 100 will still be the correct size. When you destroy the pool the processes 101-110 go away. Process 100 remains and is still the correct size. When you create a new pool the whole thing repeats with 10 new process IDs.

Yep. You are right. thanks.